service/src/test/java/bio/terra/workspace/service/workspace/flight/WorkspaceDeleteFlightTest.java

```diff
@@ -0,0 +1,155 @@
+
+    FlightState flightState =
+        StairwayTestUtils.blockUntilFlightCompletes(
+            jobService.getStairway(),
+            WorkspaceDeleteFlight.class,
+            deleteParameters,
+            DELETION_FLIGHT_TIMEOUT,
+            debugInfo);
+    assertEquals(FlightStatus.SUCCESS, flightState.getFlightStatus());
+
+    // Verify the resource and workspace are not in WSM DB
+    assertThrows(
+        WorkspaceNotFoundException.class,
+        () ->
+            controlledResourceService.getControlledResource(
+                dataset.getWorkspaceId(), dataset.getResourceId(), userReq));
+    assertThrows(
+        WorkspaceNotFoundException.class,
+        () -> workspaceService.getWorkspace(workspace.getWorkspaceId(), userReq));
+  }
+
+  @Test
+  @DisabledIfEnvironmentVariable(named = "TEST_ENV", matches = BUFFER_SERVICE_DISABLED_ENVS_REG_EX)
+  void cannotUndoWorkspaceDelete() throws Exception {
+    // Create a workspace with a controlled resource
+    AuthenticatedUserRequest userReq = userAccessUtils.defaultUserAuthRequest();
+    Workspace workspace = connectedTestUtils.createWorkspaceWithGcpContext(userReq);
+    ControlledBigQueryDatasetResource dataset =
+        ControlledResourceFixtures.makeDefaultControlledBigQueryDatasetResource()
+            .workspaceId(workspace.getWorkspaceId())
+            .build();
+    var creationParameters =
+        ControlledResourceFixtures.defaultBigQueryDatasetCreationParameters()
+            .datasetId(dataset.getDatasetName());
+    controlledResourceService.createBqDataset(
+        dataset, creationParameters, Collections.emptyList(), userReq);
+
+    ControlledResource gotResource =
+        controlledResourceService.getControlledResource(
+            workspace.getWorkspaceId(), dataset.getResourceId(), userReq);
+    assertEquals(dataset, gotResource.castToBigQueryDatasetResource());
+
+    FlightMap deleteParameters = new FlightMap();
+    deleteParameters.put(
+        WorkspaceFlightMapKeys.WORKSPACE_ID, workspace.getWorkspaceId().toString());
+    deleteParameters.put(
+        WorkspaceFlightMapKeys.WORKSPACE_STAGE, workspace.getWorkspaceStage().name());
+    deleteParameters.put(JobMapKeys.AUTH_USER_INFO.getKeyName(), userReq);
+
+    // Deletion steps can't be undone, so if the flight fails at the end, this should be marked as
+    // a dismal failure and the deletion should persist.
+    FlightDebugInfo debugInfo = FlightDebugInfo.newBuilder().lastStepFailure(true).build();
+    FlightState flightState =
+        StairwayTestUtils.blockUntilFlightCompletes(
+            jobService.getStairway(),
+            WorkspaceDeleteFlight.class,
+            deleteParameters,
+            DELETION_FLIGHT_TIMEOUT,
+            debugInfo);
+    assertEquals(FlightStatus.FATAL, flightState.getFlightStatus());
+
+    // Verify the resource and workspace are still deleted, as delete steps have no undo.
+    assertThrows(
+        WorkspaceNotFoundException.class,
```

Comment:
shouldn't this be `ResourceNotFoundException`?

Reply:
If you ask for a resource that doesn't exist in a workspace that also doesn't exist, we throw `WorkspaceNotFound`. You'll only get `ResourceNotFound` if you ask for a resource that doesn't exist in a workspace that does

Reply:
that makes sense